forch/faucetizer.py

```diff
@@ -1,49 +1,164 @@
+        self._faucet_config = None
+        self._output_file = output_file
 
     def process_network_state(self, network_state):
-        """Process network state inputs"""
-        self._network_state = proto_dict(network_state, True)
+        """Process network state input"""
+        macs = set()
+        for mac, learning in network_state.device_mac_learnings.items():
+            macs.add(mac)
+            self._process_device_learning(mac, learning)
+        for mac, behavior in network_state.device_mac_behaviors.items():
+            macs.add(mac)
+            self._process_device_behavior(mac, behavior)
 
-        sorted_rules = list(self._network_state.get("named_match_rules", {}).keys())
-        sorted_rules.sort()
-        sys.stdout.write(f'{sorted_rules}\n')
+        self._faucetize(macs)
 
-        sorted_macs = list(self._network_state.get('device_mac_behaviors', {}).keys())
-        sorted_macs.sort()
-        sys.stdout.write(f'{sorted_macs}\n')
+    def _process_device_learning(self, mac, learning):
+        device_map = self._devices.setdefault(mac, {})
+        device_map['new_switch'] = learning.switch
+        device_map['new_port'] = learning.port
 
+    def _process_device_behavior(self, mac, behavior):
+        device_map = self._devices.setdefault(mac, {})
+        device_map['vlan'] = behavior.vid
+        device_map['role'] = behavior.role
 
-def load_network_state(base_dir_name):
+    def process_faucet_config(self, faucet_config):
+        """Process faucet config when base faucet config changes"""
+        self._base_faucet_config = faucet_config
+        self._faucet_config = faucet_config
+        self._faucetize()
+
+    def _faucetize(self, macs=None):
+        if not self._faucet_config:
+            return
+
+        if not macs:
+            macs = self._devices.keys()
+
+        commit = False
+
+        for mac in macs:
+            old_switch = self._devices.get(mac, {}).get('old_switch')
+            old_port = self._devices.get(mac, {}).get('old_port')
+            new_switch = self._devices.get(mac, {}).get('new_switch')
+            new_port = self._devices.get(mac, {}).get('new_port')
+
+            if old_switch != new_switch or old_port != new_port:
+                self._reset_port_config(old_switch, old_port)
+                commit = True
+
+            vlan = self._devices.get(mac, {}).get('vlan')
+            role = self._devices.get(mac, {}).get('role')
+
+            if new_switch and vlan and role:
+                switch_config = self._faucet_config.get('dps', {}).get(new_switch, {})
+                port_config = switch_config.get('interfaces', {}).get(new_port)
+                if not port_config:
+                    LOGGER.warning('Switch or port not defined in faucet config: %s %s',
+                                   new_switch, new_port)
+                    continue
+                port_config['native_vlan'] = vlan
+                port_config['acls_in'] = [f'role_{role}']
```

Comment:
what if role is not specified?

Reply:
This will not be handled by authenticator? Does it mean authenticator will return a device with VLAN but not role?